machinelearning/tensorflow.md

```diff
@@ -0,0 +1,415 @@
+
+```
+from tensorflow.python.tools.optimize_for_inference_lib import optimize_for_inference
+
+graph_def = tf.compat.v1.GraphDef()
+with tf.compat.v1.gfile.FastGFile(model_path, "rb") as f:
+     graph_def.ParseFromString(f.read())
+
+optimized_graph_def = optimize_for_inference(graph_def, [item.split(':')[0] for item in inputs],
+                    [item.split(':')[0] for item in outputs], dtypes.float32.as_datatype_enum, False)
+g = tf.compat.v1.import_graph_def(optimized_graph_def, name='')
+```
+Note: While the Grappler optimizer covers majority of the networks, there are few scenarios where either the Grappler optimizer can't optimize the generic graph or the runtime kernel launch overhead is simply not acceptable. XLA addresses these gaps by providing an alternative mode of running models: it compiles the TensorFlow graph into a sequence of computation kernels generated specifically for the given model. Please refer to the **Enable XLA optimizations** section below to compile Tensorflow with the downstream optimizations and using them for inference.
+
+# Evaluate performance with standard MLPerf inference benchmarks
+
+1. Setup MLPerf inference benchmarks and the required tools.
+
+```
+sudo apt install -y build-essential cmake libgl1-mesa-glx libglib2.0-0 libsm6 libxrender1 libxext6 python3-pip
+
+git clone https://github.com/mlcommons/inference.git --recursive
+cd inference
+git checkout 5ec3ac922556107ce0f6ca63c175d379017ba3d8
+cd loadgen
+CFLAGS="-std=c++14" python3 setup.py bdist_wheel
+pip install <dist/*.whl>
+
+```
+2. Benchmark image classification with Resnet50
+
+```
+sudo apt install python3-ck
+ck pull repo:ck-env
+
+# Download ImageNet's validation set
+# These will be installed to ${HOME}/CK_TOOLS/
+# Select option 1: val-min data set.
+ck install package --tags=image-classification,dataset,imagenet,aux
+ck install package --tags=image-classification,dataset,imagenet,val
+
+# Copy the labels into the image location
+cp ${HOME}/CK-TOOLS/dataset-imagenet-ilsvrc2012-aux/val.txt ${HOME}/CK-TOOLS/dataset-imagenet-ilsvrc2012-val-min/val_map.txt
+
+cd inference/vision/classification_and_detection
+wget https://zenodo.org/record/2535873/files/resnet50_v1.pb
+
+# install additional packages
+pip install opencv-python
+pip install pycocotools
+
+#set the data and model path
+export DATA_DIR=${HOME}/CK-TOOLS/dataset-imagenet-ilsvrc2012-val-min
+export MODEL_DIR=${HOME}/inference/vision/classification_and_detection
+
+# setup the tensorflow thread pool parameters via MLPerf env variables
+export MLPERF_NUM_INTER_THREADS=1
+export MLPERF_NUM_INTRA_THREADS= <#of vcpus>
```

Comment:
`nproc` ? instead of having the to be filled in value?

Reply:
no i mean the program 'nproc' to provide the number of cpus instead of something they have to fill in
`export MLPERF_NUM_INTRA_THREADS=$(nproc)`

Reply:
ohk, got it. thanks.  it also reminded me that I need to take care of multi process scenario.